controllers/remote/cluster_cache.go

```diff
@@ -0,0 +1,401 @@
+	m.clusterCachesLock.Lock()
+	defer m.clusterCachesLock.Unlock()
+
+	// If another goroutine created the cache while this one was waiting to acquire the write lock, return that
+	// instead of overwriting it.
+	if c, exists := m.clusterCaches[cluster]; exists {
+		return c, nil
+	}
+
+	config, err := RESTConfig(ctx, m.client, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
+	}
+
+	remoteCache, err := cache.New(config, cache.Options{Scheme: m.scheme})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating cache for remote cluster")
+	}
+	stop := make(chan struct{})
+
+	cc := &clusterCache{
+		Cache: remoteCache,
+		stop:  stop,
+	}
+	m.clusterCaches[cluster] = cc
+
+	// Start the cache!!!
+	go remoteCache.Start(cc.stop)
+	// Start cluster healthcheck!!!
+	go m.healthCheckCluster(ctx, cluster, config, healthCheckPollInterval, healthCheckRequestTimeout, healthCheckUnhealthyThreshold, "/")
+
+	return cc, nil
+}
+
+func (m *ClusterCacheTracker) deleteClusterCache(cluster client.ObjectKey) {
+	m.clusterCachesLock.Lock()
+	defer m.clusterCachesLock.Unlock()
+
+	delete(m.clusterCaches, cluster)
+}
+
+type ClusterCacheReconciler struct {
+	log     logr.Logger
+	client  client.Client
+	tracker *ClusterCacheTracker
+}
+
+// healthCheckCluster will poll the cluster's API at the path given and, if there are
+// `unhealthyThreshold` consecutive failures, will deem the cluster unhealthy.
+// Once the cluster is deemed unhealthy, the cluster's cache is stopped and removed.
+func (m *ClusterCacheTracker) healthCheckCluster(ctx context.Context, cluster client.ObjectKey, cfg *rest.Config, interval, requestTimeout time.Duration, unhealthyThreshold int, path string) {
+	unhealthyCount := 0
+
+	runHealthCheckWithThreshold := func() (bool, error) {
+		remoteCache := m.getClusterCache(cluster)
+		if remoteCache == nil {
+			// Cache for this cluster has already been cleaned up.
+			// Nothing to do, so return true.
+			return true, nil
+		}
+
+		// healthCheckPath returning an error is considered a failed health check
+		// (Either an issue was encountered connecting or the API returned an error).
+		// If no error occurs, reset the unhealthy coutner.
+		err := healthCheckPath(cfg, requestTimeout, path)
+		if err != nil {
+			unhealthyCount++
+		} else {
+			unhealthyCount = 0
+		}
+
+		if unhealthyCount >= unhealthyThreshold {
+			// `healthCheckUnhealthyThreshold` (or more) consecutive failures.
+			// Cluster is now considered unhealthy.
+			// return last error from `doHealthCheck`
+			return false, err
+		}
+
+		return false, nil
+	}
+
+	err := wait.PollImmediateUntil(interval, runHealthCheckWithThreshold, ctx.Done())
+	// An error returned implies the health check has failed a sufficient number of
+	// times for the cluster to be considered unhealthy
+	if err != nil {
+		fmt.Printf("error: %v", err)
```

Comment:
Did you mean to leave this in? If so it should be using `m.log`

Reply:
Good catch, I think this was me debugging, removed